examples/soupbintcp-client/src/main/java/com/paritytrading/nassau/soupbintcp/client/TestClient.java

```diff
@@ -8,62 +8,69 @@
 
 class TestClient {
 
-    private static final String USAGE = "nassau-soupbintcp-client <host> <port> <packets>";
+    private static final String USAGE = "nassau-soupbintcp-client <host> <port> <packets> <packets-per-second>";
 
     public static void main(String[] args) throws IOException {
-        if (args.length != 3)
+        if (args.length != 4)
             usage(USAGE);
 
         try {
-            String host   = args[0];
-            int    port   = Integer.parseInt(args[1]);
-            int    orders = Integer.parseInt(args[2]);
+            String host             = args[0];
+            int    port             = Integer.parseInt(args[1]);
+            int    packets          = Integer.parseInt(args[2]);
+            int    packetsPerSecond = Integer.parseInt(args[3]);
 
-            main(new InetSocketAddress(host, port), orders);
+            main(new InetSocketAddress(host, port), packets, packetsPerSecond);
         } catch (NumberFormatException e) {
             usage(USAGE);
         }
     }
 
-    private static void main(InetSocketAddress address, int packets) throws IOException {
-        final Client client = Client.connect(address);
+    private static void main(InetSocketAddress address, int packets, int packetsPerSecond) throws IOException {
+        try (final Client client = Client.connect(address)) {
+            ByteBuffer buffer = ByteBuffer.allocate(Long.BYTES);
 
-        ByteBuffer buffer = ByteBuffer.wrap(new byte[] { 'f', 'o', 'o' });
+            System.out.println("Warming up...");
 
-        System.out.println("Warming up...");
+            for (long sendAt = System.nanoTime(); client.receiveCount() < packets; client.receive()) {
+                if (System.nanoTime() >= sendAt) {
+                    buffer.asLongBuffer().put(sendAt);
 
-        for (int i = 0; i < packets; i++) {
-            client.send(buffer);
+                    client.send(buffer);
 
-            buffer.flip();
+                    buffer.flip();
 
-            client.receive();
-        }
+                    sendAt += 1_000_000_000 / packetsPerSecond;
+                }
+            }
```

Comment:
I think you had one interim iteration that extracted this loop into a separate method. What do you think about restoring that?

What about precalculating the `sendAt` increment? 🤓

Reply:
Implemented `performExchange` with this loop and introduced `increment` as a single calculation.